pkg/sources/sources.go

```diff
@@ -0,0 +1,249 @@
+	out := make(config.IndexableMap[config.Sources])
+	for name, t := range sources {
+		if _, ok := r.bindings[name]; ok {
+			out[name] = t
+		}
+	}
+	return out
+}
+
+func (r *Router) BuildTable(cfg *config.Config) *Router {
+	sources := r.GetBoundSources(cfg.Sources)
+	mergedEvents := mergeResourceEvents(sources)
+
+	for resource, resourceEvents := range mergedEvents {
+		eventRoutes := mergeEventRoutes(resource, sources)
+		r.buildTable(resource, resourceEvents, eventRoutes)
+	}
+	r.log.Debugf("sources routing table: %+v", r.table)
+	return r
+}
+
+func mergeResourceEvents(sources config.IndexableMap[config.Sources]) mergedEvents {
+	out := map[string]map[config.EventType]struct{}{}
+	for _, srcGroupCfg := range sources {
+		for _, resource := range srcGroupCfg.Kubernetes.Resources {
+			if _, ok := out[resource.Name]; !ok {
+				out[resource.Name] = make(map[config.EventType]struct{})
+			}
+			for _, e := range flattenEvents(resource.Events) {
+				out[resource.Name][e] = struct{}{}
+			}
+		}
+	}
+	return out
+}
+
+func mergeEventRoutes(resource string, sources config.IndexableMap[config.Sources]) map[config.EventType][]Route {
+	out := make(map[config.EventType][]Route)
+	for srcGroupName, srcGroupCfg := range sources {
+		for _, r := range srcGroupCfg.Kubernetes.Resources {
+			for _, e := range flattenEvents(r.Events) {
+				switch {
+				case resource == r.Name && e == config.UpdateEvent:
+					out[e] = append(out[e], Route{
+						source:     srcGroupName,
+						namespaces: r.Namespaces,
+						updateSetting: config.UpdateSetting{
+							Fields:      r.UpdateSetting.Fields,
+							IncludeDiff: r.UpdateSetting.IncludeDiff,
+						}})
+				case resource == r.Name && e != config.UpdateEvent:
+					out[e] = append(out[e], Route{source: srcGroupName, namespaces: r.Namespaces})
+				}
+			}
+		}
+	}
+	return out
+}
+
+func (r *Router) buildTable(resource string, events map[config.EventType]struct{}, routes map[config.EventType][]Route) {
+	for evt := range events {
+		if _, ok := r.table[resource]; !ok {
+
+			r.table[resource] = []Entry{{
+				event:  evt,
+				routes: routes[evt],
+			}}
+
+		} else {
+			r.table[resource] = append(r.table[resource], Entry{event: evt, routes: routes[evt]})
+		}
+	}
+}
+
+func (r *Router) RegisterInformers(targetEvents []config.EventType, handler registrationHandler) {
+	resources := r.resourcesForEvents(targetEvents)
+	for _, resource := range resources {
+		r.registrations[resource] = Registration{
+			informer:   handler(resource),
+			events:     r.resourceEvents(resource),
+			log:        r.log,
+			mapper:     r.mapper,
+			dynamicCli: r.dynamicCli,
+		}
+	}
+}
+
+func (r *Router) MapWithEventsInformer(srcEvent config.EventType, dstEvent config.EventType, handler registrationHandler) {
+	srcResources := r.resourcesForEvents([]config.EventType{srcEvent})
+	if len(srcResources) == 0 {
+		return
+	}
+
+	dstResource := "v1/events"
```

Comment:
I am a bit concerned about static resource names since, if there is a change to those versions between k8s cluster versions, botkube will stop working even it can handle those resources even version is changed.

Reply:
Any ideas?

Reply:
I prefer to do;
- Collect kubernetes built in resources during application startup and put them in a data structure.
- Refer to that variable to see what are the resource names
Again, this is not a release blocker, just wanted to point out for background compatibility.

Reply:
Calling K8s API client to get the api group and version of `events` would still require us to have some assumptions like resource name + API Group. And also, I believe the v1 K8s API will stay there for a long time with us, so IMHO moving it to consts to the top of this file would be enough for now 👍